Add Asuswrt config flow

ollo69 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR is to add Config Flow to AsusWRT integration.
All sensors are initially created as disabled and can be enabled from AsusWRT device instead of using `configuration.yaml`
I also added unit of measure for sensors that was missed in last AsusWRT PR.
This is the first step, after merge I will create a new PR to replace legacy `DeviceScanner` with `ScannerEntuty`

This replace PR #43839

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16012

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ollo69 committed:
Add AsusWRT Config Flow

ollo69 committed:
Fixed isort and flake8

ollo69 committed:
Fix comment

ollo69 commented:
Hey  @kennedyshead and @MartinHjelmare, any news about this PR?
I forgot something or just need to wait?

Kane610 commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+        if user_input is None:
+            user_input = {}
+
+        return self.async_show_form(
```
Its a lot of parameters for one step, is there no logical way to break this up into multiple steps?

ollo69 replied:
The only logic that I can see is that selecting Telnet as protocol SSH key does not make sense. Does make sense separate this?

Kane610 replied:
Sounds like unnecessary layer for one option

Kane610 commented on homeassistant/components/asuswrt/device_tracker.py:
```diff
@@ -3,11 +3,16 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+async def async_setup_entry(hass, entry, async_add_entities):
```
You should migrate the device tracker to the config entry friendly format, UniFi integration is one example

ollo69 replied:
I already did this, but was asked to me to do this with multiple PR, for this reason I'm merging to dedicated branch and not in dev.

ollo69 commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,120 +1,136 @@
-DEFAULT_DNSMASQ = "/var/lib/misc"
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
```
And we don't want this?

Kane610 replied:
You should set an end date for when it will be removed is what I mean

ollo69 commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,120 +1,136 @@
+        conf.pop(CONF_PUB_KEY, "")
+        conf.pop(CONF_SENSORS, {})
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
```
But abort immediately inside config flow. As I know I cannot determinate here if config entry already exist.

Kane610 replied:
You can easily check if any integration of a specific domain has been created https://github.com/home-assistant/core/blob/06967245066deb4c387c6e2f3fa06a5e0278a699/homeassistant/config_entries.py#L601

ollo69 committed:
Requested changes

ollo69 committed:
Invalidation version changea

ollo69 committed:
Furce checks run

Kane610 commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,120 +1,141 @@
-        )
-    },
+    vol.All(
+        cv.deprecated(DOMAIN, invalidation_version="0.121"),
```
They changed the version semantics to so 0.121 would be 2021.02.0

ollo69 replied:
Seems that I have to remove in this version (see PR #44156)

ollo69 committed:
Adapt invalidation_version to new semantic

ollo69 committed:
Remove invalidation_version from deprecated

- Remove invalidation_version from deprecated. We don't follow up and just hurts releases

Kane610 commented:
You need to increase coverage of config flow, before this can be accepted into dev

ollo69 commented:
> You need to increase coverage of config flow, before this can be accepted into dev

Because this PR is not going directly in dev, I do not work on coverage. BTW coverage for `config_flow.py` is about 99%, what is lowering here is `sensor.py` in error management that for sure I will change.
Anyway if you think that is required to go ahead I can fix this, but I need to add test that I have to remove with the next step,

Kane610 commented:
Personally I think it is good to have those negative scenarios covered by tests as well. We can await a second pair of eyes on feedback first if you prefer.

ollo69 commented:
It is not a big problem, I can fix this evening. If in the meantime another feedback will come is welcome.

MartinHjelmare commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,120 +1,141 @@
-        return True
+    async def async_close_connection(event):
+        """Close AsusWrt connection on HA Stop."""
+        if hasattr(api.connection, "disconnect"):
```
Why is the api dynamic?

ollo69 replied:
Because `disconnect` method is available only for `telnet` and not `ssh`.
I thought it would be good manage disconnection when required and method is available.

MartinHjelmare replied:
Do we know the type of connection from the config entry data? If so, I'd suggest to check that instead of using `hasattr`.

MartinHjelmare commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,120 +1,141 @@
-    if not api.is_connected:
-        _LOGGER.error("Error connecting %s to %s", DOMAIN, conf[CONF_HOST])
-        return False
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_close_connection)
```
We should save the return value here and use it to unsubscribe the listener on entry unload.

MartinHjelmare commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            _LOGGER.error("Error connecting to the AsusWrt router at %s", self._host)
+            errors["base"] = "cannot_connect"
+
+        except OSError:
```
Please only wrap the line that can raise in the try... except block. Use an `else:` block if needed.

MartinHjelmare commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            )
+            errors["base"] = "unknown"
+
+        return self._show_setup_form(user_input, errors)
```
Just check the connection and return the validated data on success and raise an error on error. Don't create the entry and don't show the form from this helper.

MartinHjelmare replied:
See our scaffold script for an example.

MartinHjelmare commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            return self._show_setup_form(user_input, errors)
+
+        self._host = host
+        self._name = user_input.get(CONF_NAME, host)
```
We don't need to ask for name, I think.

ollo69 replied:
The idea was to have a more clear name, because host name should be an IP, but that's ok.

Related to the single config entry, this is not fully true. Let me say that I should be able to manage multiple config entry but I need to implement the part related to the `scanner_entity`, that we choose to introduce as second PR.
But the only unique_id that I can use is the IP Address and as you stated in next comment it is not allowed.

MartinHjelmare replied:
Ok.

We can use the config entry id as unique id I think. It's not limited to a single entry, I was thinking of DOMAIN as unique id in that case. I'm not sure why I wrote that.

ollo69 replied:
Sorry, little be confused here. How can I get the entry id to set it as unique id if the entry is still not created?

ollo69 replied:
I think that I cannot manage multiple entry if I don't have a real unique ID for the device (e.g. MAC ADDRESS). The HostName  and IPAddress can change, at the end it should be possible to have an entry related to an IP while it is using another one because IP is changed. Is very rare case that a router change the IP, but it could be. Same consideration can be done for hostname.
I could use a package like [this](https://pypi.org/project/getmac/0.6.0/) to determinate mac address during setup but I don't now if is allowed and make sense.

MartinHjelmare replied:
I'm asking about the entity unique_id not the config entry unique_id. The entities will be created after the config entry so we'llkknow the config entry id.

ollo69 replied:
Yes, that's fine, probably this comment is more related to other comment. I report there

MartinHjelmare commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+        self._name = user_input.get(CONF_NAME, host)
+
+        # Check if already configured
+        await self.async_set_unique_id(ip_address)
```
We don't allow setting up addresses as unique id. If there are no valid unique id we can just compare existing entry ip address and abort if the user inputs the same address.

ollo69 replied:
I don't have other information available for this. Should I give up multiple entities (see previous comment)?

MartinHjelmare replied:
We can store the ip address in the config entry data and iterate all existing entries checking for the ip address in question.

ollo69 replied:
Probably do not make sense, I cannot handle multiple ID because I don't have a real UniqueID, so is enough to abort if 1 entry exist

MartinHjelmare replied:
Ok.

MartinHjelmare commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+                    vol.Required(CONF_MODE, default="router"): vol.In(
+                        {"router": "Router", "ap": "Access Point"}
+                    ),
+                    vol.Required(CONF_REQUIRE_IP, default=True): bool,
```
Is this option required to be able to connect to the router? We only want required options for the connection in the main flow. Optional options should go in an options flow.

ollo69 replied:
All this parameters are used by API constructor, cannot be changed after router instance is created and cannot be changed at runtime.

MartinHjelmare replied:
Right, but can we connect to the router regardless of this option? We can reload the integration when changing this option in an options flow.

ollo69 replied:
In this case same consideration can be done for next 2 config option. But why don't allow the user to select the right option during setup? And how I reload integration (some example)?

MartinHjelmare replied:
As I said we only want the required settings in the main flow.

https://developers.home-assistant.io/docs/config_entries_options_flow_handler#signal-updates

Reload the config entry with `hass.config_entries.async_reload`.

ollo69 committed:
Review with requested changes

ollo69 commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -80,11 +85,17 @@ async def async_setup(hass, config):
     if DOMAIN in domains_list:
         return True
 
+    # remove not required config keys
```
That was my question, you are too fast 👍

ollo69 committed:
Import Options from config yaml when empty

ollo69 committed:
Fix pylint

ollo69 committed:
Skip yaml require_ip option for ap mode

MartinHjelmare commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,120 +1,178 @@
+    # import options from yaml if empty
+    yaml_options = hass.data.get(DOMAIN, {}).pop("yaml_options", {})
+    if not entry.options:
+        entry.options = MappingProxyType(yaml_options or {})
```
Please use `hass.config_entries.async_update_entry(

ollo69 committed:
Requested change

MartinHjelmare reviewed (approved):
Looks good!

ollo69 committed:
Remove unused import

ollo69 commented:
What is next step here? I should work on documentation or we can go ahead with next step (scanner entity)?

MartinHjelmare commented:
Please open a docs PR and reference it here. You can expand on the docs PR as we add more core PRs.

MartinHjelmare commented:
Just a side note: Linking PRs don't spawn github notifications. It's fine to ping the reviewer when a task like that is done. :+1:
